src/operator/z1.rs

```diff
@@ -110,47 +110,211 @@ where
     }
 }
 
-#[test]
-fn sum_circuit() {
-    let mut z1 = Z1::new(0);
-
-    let expected_result = vec![0, 1, 2, 0, 4, 5];
-
-    // Test `UnaryOperator` API.
-    let mut res = Vec::new();
-    z1.clock_start(0);
-    res.push(z1.eval(&1));
-    res.push(z1.eval(&2));
-    res.push(z1.eval(&3));
-    z1.clock_end(0);
-
-    z1.clock_start(0);
-    res.push(z1.eval_owned(4));
-    res.push(z1.eval_owned(5));
-    res.push(z1.eval_owned(6));
-    z1.clock_end(0);
-
-    assert_eq!(res, expected_result);
-
-    // Test `StrictUnaryOperator` API.
-    let mut res = Vec::new();
-    z1.clock_start(0);
-    res.push(z1.get_output());
-    z1.eval_strict(&1);
-    res.push(z1.get_output());
-    z1.eval_strict(&2);
-    res.push(z1.get_output());
-    z1.eval_strict(&3);
-    z1.clock_end(0);
-
-    z1.clock_start(0);
-    res.push(z1.get_output());
-    z1.eval_strict_owned(4);
-    res.push(z1.get_output());
-    z1.eval_strict_owned(5);
-    res.push(z1.get_output());
-    z1.eval_strict_owned(6);
-    z1.clock_end(0);
-
-    assert_eq!(res, expected_result);
+/// z^-1 operator over streams of streams.
+///
+/// The operator stores a complete nested stream consumed at the last iteration
+/// of the parent clock and outputs it at the next parent clock cycle
+/// It outputs a stream of zeros value in the first parent clock tick.
+///
+/// It is a strict operator.
+///
+/// # Examples
+///
+/// Input (one row per parent timestamps):
+///
+/// ```text
+/// 1 2 3 4
+/// 1 1 1 1 1
+/// 2 2 2 0 0
+/// ```
+///
+/// Output:
+///
+/// ```text
+/// 0 0 0 0
+/// 1 2 3 4 0
+/// 1 1 1 1 1
+/// ```
+pub struct Z1Nested<T> {
+    zero: T,
+    timestamp: usize,
+    val: Vec<T>,
+}
+
+impl<T> Z1Nested<T>
+where
+    T: Clone,
+{
+    pub fn new(zero: T) -> Self {
+        Self {
+            zero,
+            timestamp: 0,
+            val: vec![],
+        }
+    }
+
+    fn reset(&mut self) {
+        self.timestamp = 0;
+        self.val = vec![];
+    }
+}
+
+impl<T> Operator for Z1Nested<T>
+where
+    T: Clone + 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Z^-1 (nested)")
+    }
+
+    fn clock_start(&mut self, scope: Scope) {
+        if scope == 0 {
+            self.val.truncate(self.timestamp);
+        }
+        self.timestamp = 0;
+    }
+
+    fn clock_end(&mut self, scope: Scope) {
+        if scope > 0 {
+            self.reset();
+        }
+    }
+}
+
+impl<T> StrictOperator<T> for Z1Nested<T>
+where
+    T: Clone + 'static,
+{
+    fn get_output(&mut self) -> T {
+        if self.timestamp >= self.val.len() {
+            assert_eq!(self.timestamp, self.val.len());
+            self.val.push(self.zero.clone());
```

Comment:
If this is sparse there may be a better way to represent it.

Reply:
Let's not optimize prematurely.  My current assumption is there won't be many nested iterations (at least not enough to worry about the memory used by zeros).